jena-tdb/src/main/java/org/apache/jena/tdb/transaction/TransactionManager.java

```diff
@@ -436,8 +469,96 @@ public void notifyAbort(Transaction transaction)
         switch ( transaction.getMode() )
         {
             case READ: break ;
-            case WRITE: writersWaiting.release() ;
+            case WRITE: releaseWriterLock();
+        }
+    }
+    
+    private void releaseWriterLock() {
+        int x = writersWaiting.availablePermits() ;
+        if ( x != 0 )
+            throw new TDBTransactionException("TransactionCoordinator: Probably mismatch of enable/disableWriter calls") ;
+        writersWaiting.release() ;
+    }
+    
+    private boolean acquireWriterLock(boolean canBlock) {
+        if ( ! canBlock )
+            return writersWaiting.tryAcquire() ;
+        try { 
+            writersWaiting.acquire() ; 
+            return true;
+        } catch (InterruptedException e) { throw new TDBTransactionException(e) ; }
+    }
+    
+    /** Block until no writers are active.
```

Comment:
Just a "wordsmithing" suggestion: this might be better as `waitForWriters` or the like. `disableWriters` sounds a bit like the caller is going to have some kind of actual impact on the outstanding writers, but in truth the caller is just waiting for them to finish on their own, right? Same for the next method.